woocommerce/woocommerce-android
# Handle EAN-13, EAN-8 barcode formats

Closes: #9238 


### Description

This PR implements a logic to handle EAN-13 and EAN-8 barcode formats.

Whenever the barcode is scanned successfully and the product search API fails for any reason, we check the barcode format. If it's either EAN-13, EAN-8, UPC-A, or UPC-E we re-try the product search API with the newly calculated SKU. The new SKU is calculated by removing the check digit from the originally obtained SKU value.

More context: pdfdoF-2Zq-p2

<img width="600" alt="Screenshot 2023-06-05 at 9 56 17 AM" src="https://github.com/woocommerce/woocommerce-android/assets/1331230/f7ea2d38-1853-4c95-acaf-4f7de0eff279">

### Testing instructions

#### Test case 1
1. Add 2 products where one product's SKU includes EAN-13 format with the check digit and the other does not. Example - `9780201379624` and `978020137962`
2.  Generate the EAN-13 barcode for `978020137962` online. Notice the barcode appends a check digit for the SKU. The SKU on the generated barcode will be `9780201379624` (https://barcode.tec-it.com/en/EAN13?data=978020137962)
3. Scan the barcode from either the order listing or the order creation screen.
4. Ensure the product with the SKU `9780201379624` gets added.

#### Test case 2
1. Add 1 product with EAN-13 generated SKU `978020137962`
2. Generate the EAN-13 barcode for `978020137962` online. Notice the barcode appends a check digit for the SKU. The SKU on the generated barcode will be `9780201379624` (https://barcode.tec-it.com/en/EAN13?data=978020137962)
3. Scan the barcode from either the order listing or the order creation screen.
4. Ensure the product with the SKU `978020137962` gets added.

#### Test case 3
1. Add 2 products where one product's SKU includes EAN-8 format with the check digit and the other does not. Example - `90311017` and `9031101`
2.  Generate the EAN-8 barcode for `9031101` online. Notice the barcode appends a check digit for the SKU. The SKU on the generated barcode will be `90311017` (https://barcode.tec-it.com/en/EAN8?data=9031101)
3. Scan the barcode from either the order listing or the order creation screen.
4. Ensure the product with the SKU `90311017` gets added.

#### Test case 4
1. Add 1 product with EAN-8 generated SKU `9031101`
2. Generate the EAN-13 barcode for `9031101` online. Notice the barcode appends a check digit for the SKU. The SKU on the generated barcode will be `90311017` (https://barcode.tec-it.com/en/EAN8?data=9031101)
3. Scan the barcode from either the order listing or the order creation screen.
4. Ensure the product with the SKU `9031101` gets added.




### Images/gif
